File: src/main/java/me/mattstudios/triumphpets/pet/components/PetMemory.java
package me.mattstudios.triumphpets.pet.components;

import me.mattstudios.triumphpets.TriumphPets;
import me.mattstudios.triumphpets.files.Configs;
import org.bukkit.Bukkit;
import org.bukkit.entity.Item;

import java.util.ArrayList;
import java.util.List;

public class PetMemory {

    private TriumphPets plugin;

    private boolean tracking;

    private List<Item> forgetList;
    private List<Item> personalBlackList;

    public PetMemory(TriumphPets plugin) {
        this.plugin = plugin;
        tracking = false;

        forgetList = new ArrayList<>();
        personalBlackList = new ArrayList<>();

        periodicallyClearForget();
    }

    public boolean isTracking() {
        return tracking;
    }

    public void setTracking(boolean tracking) {
        this.tracking = tracking;
    }

    public List<Item> getForgetList() {
        return forgetList;
    }

    public List<Item> getPersonalBlackList() {
        return personalBlackList;
    }

    private void periodicallyClearForget() {
        Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, () -> forgetList.clear(), plugin.getConfig().getInt(Configs.FORGET_LIST_TIME) * 20L, plugin.getConfig().getInt(Configs.FORGET_LIST_TIME) * 20L);
    }
}

File: common/src/main/kotlin/me/mattstudios/triumphpets/pet/components/PetMemory.kt
package me.mattstudios.triumphpets.pet.components

import me.mattstudios.mattcore.MattPlugin
import me.mattstudios.triumphpets.config.pet.PetConfig
import me.mattstudios.triumphpets.config.pet.PetProperty
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Item


/**
 * @author Matt
 */
class PetMemory(
        private val plugin: MattPlugin, val petConfig: PetConfig,
        var filterType: FilterType, val petExperience: PetExperience) {

    constructor(plugin: MattPlugin, petConfig: PetConfig, filterType: FilterType) :
            this(plugin, petConfig, filterType, PetExperience(0))

    var isTracking = false

    private val forgetTime = petConfig[PetProperty.FORGET_LIST_TIME]

    private val forgetList = mutableListOf<Item>()
    private val filteredItems = mutableSetOf<Material>()

    /**
     * Makes it so it can clear the forget list periodically
     */
    init {
        periodicallyClearForget()
    }

    /**
     * Forgets an item
     */
    fun forgetItem(item: Item) {
        forgetList.add(item)
    }

    /**
     * Checks if the item in question is forgotten or not
     */
    fun isForgotten(item: Item?): Boolean {
        return if (item == null) true else forgetList.contains(item)
    }

    /**
     * Filters the material
     */
    fun filter(material: Material) {
        filteredItems.add(material)
    }

    /**
     * Unfilters the material
     */
    fun unFilter(material: Material) {
        filteredItems.remove(material)
    }

    /**
     * Checks if the material is filtered
     */
    fun isFiltered(material: Material): Boolean {
        if (filteredItems.isEmpty()) return false
        return if (filterType == FilterType.BLACK_LIST) filteredItems.contains(material) else !filteredItems.contains(material)
    }

    /**
     * Change the filtering type
     */
    fun toggleFilterType() {
        filterType = if (filterType == FilterType.BLACK_LIST) FilterType.WHITE_LIST else FilterType.BLACK_LIST
    }

    /**
     * Every 15 minutes clears the forgotten list
     */
    private fun periodicallyClearForget() {
        Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, Runnable { forgetList.clear() }, forgetTime * 20L, forgetTime * 20L)
    }

}